### DISCUSSION TITLE
Uso de Factory

### DISCUSSION THREAD

**Discussion by lannodev** [Flutterando/forum#26] [Perguntas e Respostas] [answered]
Olá pessoal,
Eu tenho uma classe controller que extende de ValueNotifier que controla o estado da aplicação:

```dart
class WeSlideController extends ValueNotifier<bool> {
  /// WeSlideController Construction
  // ignore: avoid_positional_boolean_parameters
  WeSlideController([bool initial = false]) : super(initial);

  /// show WeSlide Panel
  void show() => value = true;

  /// hide WeSlide Panel
  void hide() => value = false;

  /// Returns if the WeSlide Panel is opened or not
  bool get isOpened => value;

  /// Create a controller to be used as footer controller
  factory WeSlideController.footer() {
    return WeSlideController(true);
  }
}
```
Porém a mesma classe irá fazer o controle de dois objetos, um painel e um footer.
a minha duvida surgiu após esse PR: https://github.com/luciano-work/we_slide/pull/21

Gostaria de saber a opinião de vocês se existe a necessidade de utilizar uma `factory` ou apenas passar o valor através de parâmetro,
segundo o criador do PR ele alega que é estranho declarar dois objetos idênticos ao mesmo tempo:

Com factory
```dart
final _controller = WeSlideController();
final _footerController = WeSlideController.footer();
```
Sem Factory
```dart
final _controller = WeSlideController();
final _footerController = WeSlideController(false);
```
Qual a opinião de vocês ?

**Comment by renanzdm** [comment]
Em minha opinião nesse caso tanto faz, pelo fato de ser um uso muito simples. Acredito que em um caso que você precisasse realizar algumas validações mais sofisticadas o uso do factory seria mais adequado. PS: Minha opinião

**Comment by kmartins** [comment, marked as answer]
Não concordo com relação a declarar dois objetos idênticos ao mesmo tempo, tipo então seria estranho quando tenho vários `TextEditingController`.

Agora com relação à leitura, com `factory` fica melhor, pode ser usado um `construtor nomeado` nesse caso também ou até mesmo `parâmetro nomeado`.

**Comment by brunosemfio** [comment]
Acho que parâmetro nomeado nesse caso é mais legível.

```
final _controller = WeSlideController();
final _footerController = WeSlideController(initial: false);
```

**Comment by mateusfccp** [comment]
É desaconselhável usar parâmetros  booleanos não nomeados. Em vez disso, use parâmetros nomeados ou faça com `factory`.

A decisão disso tem mais a ver com exaustividade. Algumas APIs excluem outras, e nesse caso é aconselhável usar factory para impedir que o usuário gere um caso inconsistente através dos parâmetros, algo que vai, necessariamente, causar um _bug_, erro em _runtime_ ou, na melhor das hipóteses, confusão.

No caso em que não temos APIs excludentes, eu usaria o parâmetro nomeado.

**Comment by lannodev** [comment]
Muito obrigado pessoal.
todas as respostas foram boas!
